src/Microsoft.ML.FastTree/TreeEnsembleFeaturizer.cs

```diff
@@ -193,19 +93,50 @@ private void IsNormalized(int iinfo, ref bool dst)
                 // An indicator vector with length = the total number of leaves in the ensemble, indicating which leaf the example
                 // ends up in all the trees in the ensemble.
-                var leafIdType = new VectorType(NumberType.Float, _owner._totalLeafCount);
+                var leafIdType = new VectorType(NumberType.Float, owner._totalLeafCount);
                 // An indicator vector with length = the total number of nodes in the ensemble, indicating the nodes on
                 // the paths of the example in all the trees in the ensemble.
                 // The total number of nodes in a binary tree is equal to the number of internal nodes + the number of leaf nodes,
                 // and it is also equal to the number of children of internal nodes (which is 2 * the number of internal nodes)
                 // plus one (since the root node is not a child of any node). So we have #internal + #leaf = 2*(#internal) + 1,
                 // which means that #internal = #leaf - 1.
                 // Therefore, the number of internal nodes in the ensemble is #leaf - #trees.
-                var pathIdType = new VectorType(NumberType.Float, _owner._totalLeafCount - _owner._ensemble.TrainedEnsemble.NumTrees);
-                OutputSchema = Schema.Create(new SchemaImpl(ectx, owner, treeValueType, leafIdType, pathIdType));
+                var pathIdType = new VectorType(NumberType.Float, owner._totalLeafCount - owner._ensemble.TrainedEnsemble.NumTrees);
+
+                // Start creating output schema with types derived above.
+                var schemaBuilder = new SchemaBuilder();
+
+                // Metadata of tree values.
+                var treeIdMetadataBuilder = new MetadataBuilder();
+                ValueGetter<VBuffer<ReadOnlyMemory<char>>> treeIdMetadataGetter = (ref VBuffer<ReadOnlyMemory<char>> value) => owner.GetTreeSlotNames(ref value);
+                treeIdMetadataBuilder.Add(MetadataUtils.Kinds.SlotNames, MetadataUtils.GetNamesType(treeValueType.VectorSize), treeIdMetadataGetter);
+
+                // Tree values must be the first output column.
+                Contracts.Assert(TreeValuesColumnId == 0);
+                schemaBuilder.AddColumn(OutputColumnNames.Trees, treeValueType, treeIdMetadataBuilder.GetMetadata());
+
+                // Metadata of leaf IDs.
+                var leafIdMetadataBuilder = new MetadataBuilder();
+                ValueGetter<VBuffer<ReadOnlyMemory<char>>> leafIdMetadataGetter = (ref VBuffer<ReadOnlyMemory<char>> value) => owner.GetLeafSlotNames(ref value);
+                treeIdMetadataBuilder.Add(MetadataUtils.Kinds.SlotNames, MetadataUtils.GetNamesType(leafIdType.VectorSize), treeIdMetadataGetter);
+
+                // leaf IDs must be the second output column.
+                Contracts.Assert(LeafIdsColumnId == 1);
+                schemaBuilder.AddColumn(OutputColumnNames.Leaves, leafIdType, leafIdMetadataBuilder.GetMetadata());
```

Comment:
I don't see that in their present form these asserts are terribly useful, since the purpose of the check was to see that the schema's indices lined up with our expectations, and this check no longer does that. What would be useful perhaps is that once the schema is created at the end of this constructor where we've assigned output schema, we can do things like this:

So three asserts along these lines might be useful:

```csharp
Contracts.Assert(OutputSchema[OutputColumnNames.Leaves].Index == LeafIdsColumnId)
```

But the straight transliteration of the old asserts is no longer serving the purpose it once did.

Reply:
I see. We have
```csharp
                // Tree values must be the first output column.
                Contracts.Assert(OutputSchema[OutputColumnNames.Trees].Index == TreeValuesColumnId);
                // leaf IDs must be the second output column.
                Contracts.Assert(OutputSchema[OutputColumnNames.Leaves].Index == LeafIdsColumnId);
                // Path IDs must be the third output column.
                Contracts.Assert(OutputSchema[OutputColumnNames.Paths].Index == PathIdsColumnId);
```
now. Many thanks!

---
In reply to: [247681048](https://github.com/dotnet/machinelearning/pull/2132#discussion_r247681048) [](ancestors = 247681048)